disk/disk.lisp

```diff
@@ -31,56 +28,101 @@
                                    :area :wired)))
     (multiple-value-bind (full-transfers partial-transfer)
         (truncate n-sectors sectors-per-4K)
-      (loop
-         for addr = lba then (+ addr sectors-per-4k)
-         for base-offset = offset then (+ base-offset 4096)
-         repeat full-transfers
-         do (read-sector disk addr sectors-per-4K wired-buffer buffer base-offset)
-         finally
-         (when (/= partial-transfer 0)
-           (read-sector disk addr partial-transfer wired-buffer buffer base-offset
-                        (+ base-offset (* partial-transfer sector-size))))))))
+      (loop :for addr := lba :then (+ addr sectors-per-4k)
+            :for base-offset := offset :then (+ base-offset 4096)
+            :repeat full-transfers
+            :do (funcall fn disk addr sectors-per-4K wired-buffer buffer base-offset)
+            :finally
+            (when (/= partial-transfer 0)
+              (funcall fn disk addr partial-transfer wired-buffer buffer base-offset
+                       (+ base-offset (* partial-transfer sector-size))))))))
+
+;;======================================================================
+;;    API
+;;======================================================================
+
+(defgeneric block-device-sector-size (disk))
+
+(defgeneric block-device-n-sectors (disk))
+
+(defgeneric block-device-read (device lba n-sectors buffer &key offset))
 
 (defgeneric block-device-write (device lba n-sectors buffer &key offset))
 
-(defun write-sector (disk addr n-sectors wired-buffer buffer base-offset &optional (end2 nil))
-  (replace wired-buffer buffer :start2 base-offset :end2 end2)
-  (multiple-value-bind (success-p error-reason)
-      (sup:disk-write disk addr n-sectors wired-buffer)
-    (unless success-p
-      (error "Disk write error: ~A" error-reason))))
+(defgeneric block-device-read-wired (device lba n-sectors wired-buffer))
```

Comment:
What is this for? Block-device-read/-write seem sufficient. The wired area should be exposed in as few places as possible and functions should try to hide it as much as possible.